plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -540,44 +540,43 @@ fu_firehose_updater_send_program_file(FuFirehoseUpdater *self,
 				      gsize sector_size,
 				      GError **error)
 {
-	g_autoptr(FuChunk) chk_last = NULL;
 	g_autoptr(FuChunkArray) chunks = fu_chunk_array_new_from_bytes(program_file,
 								       FU_CHUNK_ADDR_OFFSET_NONE,
 								       FU_CHUNK_PAGESZ_NONE,
 								       payload_size);
 
-	/* last block needs to be padded to the next sector_size,
-	 * so that we always send full sectors */
-	chk_last = fu_chunk_array_index(chunks, fu_chunk_array_length(chunks) - 1, error);
-	if (chk_last == NULL)
-		return FALSE;
-	if ((fu_chunk_get_data_sz(chk_last) % sector_size) != 0) {
-		g_autoptr(GBytes) padded_bytes = NULL;
-		gsize padded_sz = sector_size * (fu_chunk_get_data_sz(chk_last) / sector_size + 1);
-
-		padded_bytes = fu_bytes_pad(fu_chunk_get_bytes(chk_last), padded_sz);
-		fu_chunk_set_bytes(chk_last, padded_bytes);
-
-		g_return_val_if_fail(fu_chunk_get_data_sz(chk_last) == padded_sz, FALSE);
-	}
 	for (guint i = 0; i < fu_chunk_array_length(chunks); i++) {
 		g_autoptr(FuChunk) chk = NULL;
+		g_autoptr(GBytes) block_padded = NULL;
 
 		/* prepare chunk */
 		chk = fu_chunk_array_index(chunks, i, error);
 		if (chk == NULL)
 			return FALSE;
 
+		/* last block needs to be padded to the next sector_size,
+		 * so that we always send full sectors */
+		if (i == (fu_chunk_array_length(chunks) - 1) &&
+		    (fu_chunk_get_data_sz(chk) % sector_size) != 0) {
+			gsize padded_sz =
+			    sector_size * (fu_chunk_get_data_sz(chk) / sector_size + 1);
+
+			block_padded = fu_bytes_pad(fu_chunk_get_bytes(chk), padded_sz);
```

Comment:
fu_chunk_get_bytes() needs to be freed; and I think if you call fu_bytes_pad() on something that's the right size already it's fine.

Reply:
You're right, `fu_bytes_pad` can always be called here.